### PR TITLE
feat(Ranger): Pull policies from the Ranger Service and update using resources policies

### PR DESCRIPTION
**Pull request by WHBANG**
https://github.com/apache/incubator-pegasus/issues/1054

This patch implements how to pull policies from the Ranger Service
and dump policies to remote storage.

- Pull policies in JSON format from Ranger service and parse
  policies from JSON formated string.
- Create the path to save policies in remote storage, and update
  using resources policies.
- Dump policies to remote storage.
- Sync policies to app envs.
- Update the cached global/database resources policies.

### CODE REVIEW

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 1]
```diff
@@ -171,5 +212,362 @@ void ranger_resource_policy_manager::parse_policies_from_json(const rapidjson::V
...
+            table_name = app.app_name;
+        } else {
+            table_name = app.app_name.substr(database_name.size());
+        }
+
+        auto req = dsn::make_unique<dsn::replication::configuration_update_app_env_request>();
+        req->__set_app_name(app.app_name);
+        req->__set_keys(
+            {dsn::replication::replica_envs::REPLICA_ACCESS_CONTROLLER_RANGER_POLICIES});
+        bool has_match_policy = false;
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:528]: rename to `is_policy_matched` ?

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 2]
```diff
@@ -226,9 +229,12 @@ dsn::error_code ranger_resource_policy_manager::update_policies_from_ranger_serv
         // for the newly created table, its app envs must be empty. This needs to be executed
         // periodically to update the table's app envs, regardless of whether the Ranger policy is
         // updated or not.
-        CHECK_EQ_MSG(dsn::ERR_OK, sync_policies_to_app_envs(), "Sync policies to app envs failed.");
-        LOG_DEBUG("Sync policies to app envs succeeded.");
-        return dsn::ERR_OK;
+        err_code = sync_policies_to_app_envs();
+        if (err_code == dsn::ERR_OK) {
+            LOG_DEBUG("Sync policies to app envs succeeded.");
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:234]: If move this log into `sync_policies_to_app_envs()`, you can simply call `ERR_LOG_AND_RETURN_NOT_OK` directly.

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 3]
```diff
@@ -459,26 +467,32 @@ void ranger_resource_policy_manager::dump_and_sync_policies()
     update_cached_policies();
     LOG_DEBUG("Update using resources policies succeed.");
 
-    CHECK_EQ_MSG(dsn::ERR_OK, sync_policies_to_app_envs(), "Sync policies to app envs failed.");
-    LOG_DEBUG("Sync policies to app envs succeeded.");
+    if (dsn::ERR_OK == sync_policies_to_app_envs()) {
+        LOG_ERROR("Sync policies to app envs failed.");
+    } else {
+        LOG_DEBUG("Sync policies to app envs succeeded.");
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:473]: How about move this log into `sync_policies_to_app_envs()`?

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 4]
```diff
@@ -459,26 +467,32 @@ void ranger_resource_policy_manager::dump_and_sync_policies()
...
                 return;
             }
-            CHECK_EQ_MSG(e, dsn::ERR_TIMEOUT, "Dump Ranger policies to remote storage failed.");
-            LOG_ERROR("Dump Ranger policies to remote storage timeout, retry later.");
-            dsn::tasking::enqueue(LPC_CM_GET_RANGER_POLICY,
-                                  &_tracker,
-                                  [this]() { dump_policies_to_remote_storage(); },
-                                  0,
-                                  load_ranger_policy_retry_delay_ms);
+            if (e == dsn::ERR_TIMEOUT) {
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:486]: Keep to assert here, as how most zookeeper failure handled now in Pegasus.

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 5]
```diff
@@ -171,5 +215,382 @@ void ranger_resource_policy_manager::parse_policies_from_json(const rapidjson::V
...
+    ERR_LOG_AND_RETURN_NOT_OK(list_resp.err, "list_apps failed.");
+    for (const auto &app : list_resp.infos) {
+        std::string database_name = get_database_name_from_app_name(app.app_name);
+        std::string table_name;
+        if (database_name.empty()) {
+            database_name = "*";
+            table_name = app.app_name;
+        } else {
+            table_name = app.app_name.substr(database_name.size());
+        }
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:536]:
```suggestion
        std::string table_name = get_table_name_from_app_name(app.app_name);
        if (database_name.empty()) {
            database_name = "*";
        }
```

Comment (acelyc111) [reply src/runtime/ranger/ranger_resource_policy_manager.cpp:536]: If `get_database_name_from_app_name` and `get_table_name_from_app_name` are always apear in pair, consider to encapsulate a function like `std::pair<std::string, std::string> get_db_table_name_from_app_name()` instead ?

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 6]
```diff
@@ -171,5 +215,382 @@ void ranger_resource_policy_manager::parse_policies_from_json(const rapidjson::V
...
+    if (_local_policy_version > remote_policy_version) {
+        LOG_WARNING("Local Ranger policy version ({}) is larger than remote version ({}), please "
+                    "check Ranger services ({}).",
+                    _local_policy_version,
+                    remote_policy_version,
+                    FLAGS_ranger_service_name);
+        return dsn::ERR_RANGER_POLICIES_NO_NEED_UPDATE;
+    }
+
+    if (_local_policy_version == 0) {
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:346]: `_local_policy_version` should be updated if it's less than `remote_policy_version`, regardless if it equals to 0 or not?

File: src/runtime/ranger/ranger_resource_policy_manager.cpp (hunk 5 above)
Comment (WHBANG) [reply src/runtime/ranger/ranger_resource_policy_manager.cpp:536]: Keep it separate, just get the `database_name` at the database level ACL.

File: src/runtime/ranger/ranger_resource_policy_manager.cpp [hunk 7]
```diff
@@ -171,5 +215,368 @@ void ranger_resource_policy_manager::parse_policies_from_json(const rapidjson::V
...
+{
+    LOG_DEBUG("Start to sync Ranger policies to remote storage.");
+
+    dump_policies_to_remote_storage();
+    LOG_DEBUG("Dump Ranger policies to remote storage succeed.");
+
+    update_cached_policies();
+    LOG_DEBUG("Update using resources policies succeed.");
+
+    if (dsn::ERR_OK == sync_policies_to_app_envs()) {
```
Comment (acelyc111) [inline src/runtime/ranger/ranger_resource_policy_manager.cpp:464]:
```suggestion
    if (dsn::ERR_OK != sync_policies_to_app_envs()) {
```

**Review by acelyc111 — DISMISSED** [review]

**Review by acelyc111 — APPROVED** [review]

**Review by foreverneverer — APPROVED** [review]

**Pull request merged**